### PR TITLE
tools: new tools directory for cataloging tools

### LINKED ISSUE #1608: Authorlist transformation utility
@kaplun commented on [Tue Sep 27 2016](https://github.com/inspirehep/inspire/issues/217)

[...]
We often need to format lists of authors in the correct MARC format when we catalogue a new publication or to add some missing authors in Bibedit-Text-MARC. It would be nice to have a small tool that we could use easily on our PCs. Here are some examples that you could use for testing, in principle we have two different models:
- One without affiliations

```
K. Berkelman, D. Cords, R. Felst, E. Gadermann, G. Grindhammer, H. Hultschig, P. Joos, W. Koch, U. Kötz, H. Krehbiel, D. Kreinick, J. Ludwig, K.-H. Mess, K.C. Moffeit, A. Petersen, G. Poelz, J. Ringel, K. Sauerberg, P. Schmüser, G. Vogel, B.H. Wiik, G. Wolf
```

The result should be in that format:

```
100__ $$aBerkelman, K.
700__ $$aCords, D.
700__ $$aFelst, R.
700__ $$aGadermann, E.
700__ $$aGrindhammer, G.
700__ $$aHultschig, H.
700__ $$aJoos, P.
700__ $$aKoch, W.
700__ $$aKötz, U.
700__ $$aKrehbiel, H.
700__ $$aKreinick, D.
700__ $$aLudwig, J.
700__ $$aMess, K.-H.
700__ $$aMoffeit, K.C.
700__ $$aPetersen, A.
700__ $$aPoelz, G.
700__ $$aRingel, J.
700__ $$aSauerberg, K.
700__ $$aSchmüser, P.
700__ $$aVogel, G.
700__ $$aWiik, B.H.
700__ $$aWolf, G.
```
- One model with affiliations identified with numbers:

```
F. Durães1, A.V. Giannini2, V.P. Gonçalves3,4 and F.S. Navarra2
1 CERN
2 Fermilab
3 Lund University
4 Instituto de Física, Universidade de São Paulo
```

The result should be in that format :

```
100__ $$aDurães, F.$$u1
700__ $$aGiannini, A.V.$$u2
700__ $$aGonçalves, V.P.$$u3$$u4
700__ $$aNavarra, F.S$$u2
```

Or you can also replace the numbers with the subfield `__v` affiliation as it is on the list ( `700__ $$aGonçalves, V.P.$$vLund University $$vInstituto de Física, Universidade de São Paulo`)
[...]
---

@kaplun commented on [Tue Sep 27 2016](https://github.com/inspirehep/inspire/issues/217#issuecomment-249806683)

@bittirousku: @annetteholtkamp thinks this is a nice tool you might help implementing.
---

@kaplun commented on [Tue Sep 27 2016](https://github.com/inspirehep/inspire/issues/217#issuecomment-249807702)

Actually we can implement this already on Labs, and having it for the time being reachable e.g. under:

http://qa.inspirehep.net/tools/authorlist

### PR DESCRIPTION
**Pull request by bittirousku**
* Adds new authorlist tool to format author strings.

(closes #1608)

### CODE REVIEW

**Comment by bittirousku** [comment]
@kaplun @jmartinm: Any thoughts?

**Comment by kaplun** [comment]
@bittirousku have you applied all @jacquerie comments?

**Comment by kaplun** [comment]
> There's more I want to say, but since you were waiting for a review...

@jacquerie do you have any code design related comments?

**Comment by bittirousku** [comment]
@kaplun Not yet, I'm fixing a broken HEPcrawl spider at the moment.

**Comment by kaplun** [comment]
Don't worry: it's github fault: It was showing me your "any thoughts" message after @jacquerie's review.

**Comment by jacquerie** [comment]
> @jacquerie do you have any code design related comments?

Yes, and I already made a few. The main one I have right now is that the code looks vulnerable to XSS.

File: inspirehep/modules/tools/templates/authorlist_form.html [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+<form method=post action="/tools/authorlist" id="authorlist">
+  <dl>
+    {{ render_field(form.author_string) }}
+  </dl>
+  <p><input type=submit value=Format>
+</form>
+
+{% if results %}
+<h4>{{ results_title }}</h4>
+<pre><textarea class="form-control" form=authorlist id=resultarea>{{ results }}</textarea></pre>
```
Comment (kaplun) [inline inspirehep/modules/tools/templates/authorlist_form.html]: @jacquerie this is the only place where user provided input is used. Isn't this automatically escaped by Jinja2 and hence safe from XSS?

Comment (jacquerie) [inline inspirehep/modules/tools/templates/authorlist_form.html]: I did a bit of testing and yeah, you're right :+1:

File: tests/unit/tools/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, division, print_function
+
+import pytest
```
Comment (bittirousku) [inline tests/unit/tools/test_utils.py]: Actually I'm using it when testing exceptions.

File: inspirehep/modules/tools/utils.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+from __future__ import absolute_import, print_function
+
+import re
+
+from nameparser import HumanName
+import six
+
+
+def authorlist(text, affiliations=False):
```
Comment (bittirousku) [inline inspirehep/modules/tools/utils.py]: It was called before, but I removed it.

File: inspirehep/modules/tools/utils.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    Input can be in two different formats:
+        F. Lastname, F.M. Otherlastname
+    or
+        F. Lastname1, F.M. Otherlastname1,2
+        1 CERN
+        2 Otheraffiliation
+
+    If affiliations are not explicitly declared, try to detect them.
+
+    TODO: MARCXML?
```
Comment (bittirousku) [inline inspirehep/modules/tools/utils.py]: Just a question for @kaplun  et al. if we want to output also MARCXML.

File: inspirehep/modules/tools/templates/authorlist_form.html [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+## License, or (at your option) any later version.
+##
+## INSPIRE is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with INSPIRE; if not, write to the Free Software Foundation, Inc.,
+## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
```
Comment (bittirousku) [inline inspirehep/modules/tools/templates/authorlist_form.html]: Then we should also correct a lot of existing files, as I copy-pasted this from an old file.

**Comment by bittirousku** [comment]
@kaplun @jacquerie I made most of the proposed changes. Something still missing or just very wrong?

File: inspirehep/modules/tools/utils.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    return authorlist_without_affiliations(text)
+
+
+def format_name(author):
+    """Parse the raw author name and return a formatted string."""
+    name = HumanName(author)
+    lname = name.last
+    fname = name.first
+    mname = name.middle
```
Comment (bittirousku) [inline inspirehep/modules/tools/utils.py]: That's nice.

File: tests/unit/tools/test_utils.py (hunk 2 above)
Comment (jacquerie) [inline tests/unit/tools/test_utils.py]: Not when I did the first pass over your code!

File: inspirehep/modules/tools/views.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    except (AttributeError, ValueError, KeyError) as err:
+        errors = err.args
+    return render_template(
+        'authorlist_form.html',
+        results=marc,
+        title='Format author strings to MARC21',
+        form=form,
+        results_title=results_title,
+        errors=errors,
+        helptexts=helptexts
```
Comment (jacquerie) [inline inspirehep/modules/tools/views.py]: IMO the `helptexts` can be directly embedded in the template, without passing them.

File: inspirehep/modules/tools/utils.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    or
+        F. Lastname1, F.M. Otherlastname1,2
+        1 CERN
+        2 Otheraffiliation
+    """
+    if not text:
+        return ''
+    if not isinstance(text, six.text_type):
+        # FIXME: how should I handle unicode?
+        text = text.decode('utf-8')
```
Comment (jacquerie) [inline inspirehep/modules/tools/utils.py]: Here you are saying: "If this thing is not a string, convert it to utf-8", which is probably not what you meant. Maybe
```python
if not isinstance(text, unicode):
    text = text.decode('utf-8')
```

File: inspirehep/modules/tools/utils.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+    # Assume that if there are numbers in author name, they are affilation ids.
+    # Also assume that if one author has an affiliation, everyone else has too.
+    # FIXME: can we make these assumptions?
```
Comment (jacquerie) [inline inspirehep/modules/tools/utils.py]: Is this really a FIXME? This is a question about requirements, which I believe only you and the curators know...

File: inspirehep/modules/tools/utils.py (hunk 8 above)
Comment (bittirousku) [inline inspirehep/modules/tools/utils.py]: Isn't `six.text_type` == `unicode` in Python2?

File: inspirehep/modules/tools/utils.py (hunk 9 above)
Comment (bittirousku) [inline inspirehep/modules/tools/utils.py]: Well, it's a question :)

File: inspirehep/modules/tools/views.py (hunk 7 above)
Comment (bittirousku) [inline inspirehep/modules/tools/views.py]: I agree.

File: inspirehep/modules/tools/utils.py (hunk 9 above)
Comment (kaplun) [inline inspirehep/modules/tools/utils.py]: @bittirousku I think your assumptions are valid

File: inspirehep/modules/tools/utils.py (hunk 8 above)
Comment (kaplun) [inline inspirehep/modules/tools/utils.py]: @jacquerie I believe @bittirousku is right. There is no such thing as `unicode` in Python3.

**Comment by kaplun** [comment]
Merged :tada:

**Comment by david-caro** [comment]
:+1:

**Pull request closed without merging**